processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java

```diff
@@ -0,0 +1,148 @@
+import io.druid.segment.FloatColumnSelector;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.column.ValueType;
+
+import java.util.Map;
+
+public class ExpressionObjectSelector implements ObjectColumnSelector<Number>
+{
+  private final Expr expression;
+  private final Expr.ObjectBinding bindings;
+
+  private ExpressionObjectSelector(Expr.ObjectBinding bindings, Expr expression)
+  {
+    this.bindings = Preconditions.checkNotNull(bindings, "bindings");
+    this.expression = Preconditions.checkNotNull(expression, "expression");
+  }
+
+  public static ExpressionObjectSelector from(ColumnSelectorFactory columnSelectorFactory, Expr expression)
+  {
+    return new ExpressionObjectSelector(createBindings(columnSelectorFactory, expression), expression);
+  }
+
+  private static Expr.ObjectBinding createBindings(ColumnSelectorFactory columnSelectorFactory, Expr expression)
+  {
+    final Map<String, Supplier<Number>> suppliers = Maps.newHashMap();
+    for (String columnName : Parser.findRequiredBindings(expression)) {
+      final ColumnCapabilities columnCapabilities = columnSelectorFactory.getColumnCapabilities(columnName);
+      final ValueType nativeType = columnCapabilities != null ? columnCapabilities.getType() : null;
+      final Supplier<Number> supplier;
+
+      if (nativeType == ValueType.FLOAT) {
+        final FloatColumnSelector selector = columnSelectorFactory.makeFloatColumnSelector(columnName);
+        supplier = new Supplier<Number>()
+        {
+          @Override
+          public Number get()
+          {
+            return selector.get();
+          }
+        };
+      } else if (nativeType == ValueType.LONG) {
+        final LongColumnSelector selector = columnSelectorFactory.makeLongColumnSelector(columnName);
+        supplier = new Supplier<Number>()
+        {
+          @Override
+          public Number get()
+          {
+            return selector.get();
+          }
+        };
+      } else if (nativeType == null) {
+        // Unknown ValueType. Try making an Object selector and see if that gives us anything useful.
+        final ObjectColumnSelector selector = columnSelectorFactory.makeObjectColumnSelector(columnName);
+        final Class clazz = selector == null ? null : selector.classOfObject();
+        if (selector != null && (clazz == Object.class || Number.class.isAssignableFrom(clazz))) {
+          // There may be numbers here.
+          supplier = new Supplier<Number>()
+          {
+            @Override
+            public Number get()
+            {
+              return tryParse(selector.get());
+            }
+          };
+        } else {
+          // We know there are no numbers here. Use a null supplier.
+          supplier = null;
+        }
+      } else {
+        // Unhandleable ValueType (possibly STRING or COMPLEX).
```

Comment:
It _could_, and we could actually also do an impl based on DimensionSelector, but I avoided that on purpose because that would lead to expressions and field accesses behaving differently for aggregators: longSum over fieldName "x" and expression "x" would behave differently if "x" were a string column. There's a test case in SchemaEvolutionTest that verifies this and would fail if the behavior were different.